src/build.ts

```diff
@@ -0,0 +1,372 @@
+  await exec.exec('sudo', ['mkdir', workingDir, '-p'])
+  await exec.exec('sudo', ['chown', os.userInfo().username, workingDir])
+  await exec.exec('git', [
+    'clone',
+    `https://github.com/${repository}.git`,
+    '--branch',
+    ref,
+    workingDir
+  ])
+  const charmcraftSha = (
+    await exec.getExecOutput('git', ['rev-parse', 'HEAD'], { cwd: workingDir })
+  ).stdout.trim()
+  const cacheKey = `charmcraft-${charmcraftSha}`
+  const charmcraftGlob = path.join(workingDir, 'charmcraft*.snap')
+  const restored = await cache.restoreCache([charmcraftGlob], cacheKey)
+  if (!restored) {
+    await installSnapcraft()
+    core.startGroup('snapcraft pack (charmcraft)')
+    await exec.exec('snapcraft', ['--use-lxd', '--verbosity', 'trace'], {
+      cwd: workingDir
+    })
+    core.endGroup()
+  }
+  const charmcraftSnaps = await (await glob.create(charmcraftGlob)).glob()
+  if (charmcraftSnaps.length == 0) {
+    throw new Error("can't find charmcraft snap")
+  }
+  await exec.exec('sudo', [
+    'snap',
+    'install',
+    charmcraftSnaps[0],
+    '--classic',
+    '--dangerous'
+  ])
+  if (!restored) {
+    await cache.saveCache([charmcraftGlob], cacheKey)
+  }
+}
+
+interface BuildCharmParams {
+  plan: BuildPlan
+  charmcraftChannel: string
+  charmcraftRepository: string
+  charmcraftRef: string
+}
+
+async function gitTreeId(p: string): Promise<string> {
+  const gitPath = path.resolve(p) == path.resolve(process.cwd()) ? '' : p
+  const tree = (
+    await exec.getExecOutput('git', ['rev-parse', `HEAD:${gitPath}`])
+  ).stdout.trim()
+  if (!tree) {
+    throw new Error(`failed to get git tree id for path: ${p}`)
+  }
+  return tree
+}
+
+async function buildCharm(params: BuildCharmParams): Promise<void> {
+  if (params.charmcraftRepository && params.charmcraftRef) {
+    await buildInstallCharmcraft(
+      params.charmcraftRepository,
+      params.charmcraftRef
+    )
+  } else if (params.charmcraftChannel) {
+    await exec.exec('sudo', [
+      'snap',
+      'install',
+      'charmcraft',
+      '--channel',
+      params.charmcraftChannel,
+      '--classic'
+    ])
+  } else {
+    await exec.exec('sudo', ['snap', 'install', 'charmcraft', '--classic'])
+  }
+  core.startGroup('charmcraft pack')
+  await exec.exec('charmcraft', ['pack', '--verbosity', 'trace'], {
+    cwd: params.plan.source_directory
+  })
+  core.endGroup()
+  const charmFiles = await (
+    await glob.create(path.join(params.plan.source_directory, '*.charm'))
+  ).glob()
+  const artifact = new DefaultArtifactClient()
+  const manifestFile = path.join(params.plan.source_directory, 'manifest.json')
+  fs.writeFileSync(
+    manifestFile,
+    JSON.stringify(
+      { name: params.plan.name, files: charmFiles.map(f => path.basename(f)) },
+      null,
+      2
+    )
+  )
+  await artifact.uploadArtifact(
+    params.plan.output,
+    [...charmFiles, manifestFile],
+    params.plan.source_directory
+  )
```

Comment:
I'm wondering if overwrite:true is required for uploadArtifact here?

Reply:
https://github.com/actions/upload-artifact/?tab=readme-ov-file#overwriting-an-artifact

Reply:
And perhaps for other uploadArtifact calls below if this is valid

Reply:
The new `operator-workflows` will ensure each job uploads artifacts with different names, so overwrite is not required here, thanks!